src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs

```diff
@@ -454,30 +454,55 @@ AttributeLocation IAttributeTargetSymbol.AllowedAttributeLocations
             }
         }
 
+#nullable enable
         /// <summary>
         /// Symbol to copy bound attributes from, or null if the attributes are not shared among multiple source parameter symbols.
         /// </summary>
         /// <remarks>
-        /// Used for parameters of partial implementation. We bind the attributes only on the definition
-        /// part and copy them over to the implementation.
+        /// This is inconsistent with analogous 'BoundAttributesSource' on other symbols.
+        /// Usually the definition part is the source, but for parameters the implementation part is the source.
+        /// This affects the location of diagnostics among other things.
         /// </remarks>
-        private SourceParameterSymbol BoundAttributesSource
+        private SourceParameterSymbol? BoundAttributesSource
+            => PartialImplementationPart;
+
+        protected SourceParameterSymbol? PartialImplementationPart
         {
             get
             {
-                var sourceMethod = this.ContainingSymbol as SourceOrdinaryMethodSymbol;
-                if ((object)sourceMethod == null)
+                ImmutableArray<ParameterSymbol> implParameters = this.ContainingSymbol switch
+                {
+                    SourceMemberMethodSymbol method => method.PartialImplementationPart?.Parameters ?? default,
+                    SourcePropertySymbol property => property.PartialImplementationPart?.Parameters ?? default,
+                    _ => default
+                };
+
+                if (implParameters.IsDefault)
                 {
                     return null;
                 }
 
-                var impl = sourceMethod.SourcePartialImplementation;
-                if ((object)impl == null)
+                return (SourceParameterSymbol)implParameters[this.Ordinal];
```

Comment:
I wound up writing a similar assert `!this.ContainingSymbol.IsPartialImplementation()` for expediency since there isn't a helper for getting the definition part off a bare `Symbol`.